client/llb/fileop.go

```diff
@@ -749,7 +768,12 @@ func (f *FileOp) Marshal(ctx context.Context, c *Constraints) (digest.Digest, []
 			}
 		}
 
-		action, err := st.action.toProtoAction(ctx, parent, st.base)
+		// Assume that we're building an image for the same OS we're running on.
+		platform := runtime.GOOS
```

Comment:
No such assumptions in client package. This is potentially "darwin" etc.

Reply:
Hi @tonistiigi,

In the client case, I believe we wither require that the constraints which determine which platform/OS we target are set, or we leave the path alone in the client, without any type or normalization and allow the solver to take care of normalization. We can join any path elements, but the resulting path should be sent as is to the solver.

If we're running the client on Linux and building an image against a node running on Windows, with a Dockerfile targeting a Windows image, there is no way we can normalize anything in the client without knowing what platform/OS we're building for.

In the current code from `master`, we normalize the path in `toProtoc()`, which if it's not an absolute path, is joined to `/`. For windows paths this ends up turning `C:\\ProgramData` to `/C:\\ProgramData`. If we just clean it and join the `parent` to the `a.file` in `toProtoc()`, we send over something like `C:\\ProgramData/file` which is fine. We could at most do a `filepath.ToSlash()` on it to have consistent slashes like `C:/ProgramData/file`. Once it gets to the solver, it should do the right thing on both linux and Windows. If it's a relative path, it becomes absolute (joined to `/`), drive letter stripped (if it's Windows), etc.

How do you suggest we proceed?

Reply:
Gah! We still need to know if [a.file](https://github.com/moby/buildkit/blob/master/client/llb/fileop.go#LL166C39-L166C45) is absolute or relative. To do that, we need to know the OS we're targeting. A file path of `c:/test` is a valid relative path on Linux, but an absolute path on Windows. If it's relative we need to join it to the parent. If not, we return `a.file`.

Reply:
So, I looked a bit closer at this bit of code and where it ends up being called from. It seems that this is being used in the `dockerfile2llb` package. The `toDispatchState` calls the various dispatch functions which in turn end up being called by `Dockerfile2LLB()`.

`Dockerfile2LLB()` ends up being called by the `builder.Build()` in `frontend/dockerfile/builder`. This in turn is set as the `dockerfile.v0` frontend in buildkitd.

Now, when a client connects to buildkit and tries to solve a build graph, the request it sends can contain a platform or not. If we explicitly set `req.FrontendAttrs["platform"]`, the entire code path ends up having that field set. If however we don't set it, the platform constraint is nil.

We could check for a platform [in the controller](https://github.com/moby/buildkit/blob/master/control/control.go#L315) and if it's not set, we can explicitly set it to the same platform as the node the solver is running on.

Does that sound reasonable?

Reply:
Just to recap, when LLB ops (FileOp/ExecOp) are marshalled to protobuf, the paths should always be normalized to UNIX style. I think this is the case already in this PR because of the `NormalizePath` call. The `parent` in that call is also always in UNIX style because it comes from `GetDir()` which also directly gets marshaled to protobuf.

Assuming this, moving up a layer, next question is what paths do the LLB client package functions take. Do they also only take UNIX style like the protobuf they generate or do they take either Windows or UNIX path, and some platform input and then based on these values potentially convert the paths before they get marshaled to protobuf.

If former, then the client LLB pkg should not do path conversions. It may still clean up some paths but the input paths already need to be UNIX style. For Dockerfile this would mean that we determine the input platform in `dockerfile2llb/convert.go` and when we parse a path we can convert it there based on target platform. In there we can also use `runtime.GOOS` for the default platform as this is the daemon side and the default target platform is the daemon's runtime platform if none was set by the user.

Otherwise, if client LLB pkg should accept paths in multiple formats then it should be purely based on user parameters. Either there is a special way to pass it in or if it is read from the state then I think it should always default to UNIX so that there is a consistent behavior. So input paths always need to be unix unless the platform is set. In Dockerfile we need to make sure to set it always, even if the user did not pass one. We never imply platform value from `runtime.GOOS` in client pkg like it is in https://github.com/moby/buildkit/pull/3908/files#diff-0c6bf0159d4d9a07a3bae56aa2d7ee020be2009036497e0b3ef34d8a9d980869R71.

I think my preference is for the first solution but either would be acceptable.